src/lib/optimizer/strategy/predicate_pushdown_rule.cpp

```diff
@@ -13,26 +13,41 @@ namespace opossum {
 
 std::string PredicatePushdownRule::name() const { return "Predicate Pushdown Rule"; }
 
+void push_down(const std::shared_ptr<AbstractLQPNode>& node, std::shared_ptr<AbstractLQPNode> input) {
+  DebugAssert(node->left_input() && !node->right_input(), "This helper can only push down if there is a single input");
+
+  lqp_remove_node(node);
+  const auto previous_left_input = input->left_input();
+
+  input->set_left_input(node);
+  node->set_left_input(previous_left_input);
+}
+
 bool PredicatePushdownRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
   if (node->type != LQPNodeType::Predicate) return _apply_to_inputs(node);
 
   // Only predicates with exactly one output are currently supported.
   const auto outputs = node->outputs();
   if (outputs.empty() || outputs.size() > 1) return _apply_to_inputs(node);
 
-  // Find a join node
   const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(node);
 
   auto input = node->left_input();
   while (input->type == LQPNodeType::Predicate) {
+    // First, try to push down the predicates that come below. That keeps the predicate order intact.
+    if (_apply_to_inputs(node)) return true;
```

Comment:
`node` doesn't change in this loop. Maybe you meant `input`?

Anyway, I don't really get what this loop does intuitively, can you maybe try to clarify the comments a bit?